src/pytesmo/validation_framework/data_averaging.py

```diff
@@ -0,0 +1,423 @@
+            return None
+
+        if self.period is not None:
+            # here we use the isoformat since pandas slice behavior is
+            # different when using datetime objects.
+            data_df = data_df[
+                      self.period[0].isoformat():self.period[1].isoformat()]
+
+        if len(data_df) == 0:
+            warnings.warn("No data for dataset {} with arguments {:}".format(name,
+                                                                             args))
+            return None
+
+        else:
+            return data_df
+
+
+class DataAverager(MixinReadTs):
+    """
+    Class to handle multiple measurements falling under the same reference gridpoint. The goal is to include
+    here all identified upscaling methods to provide an estimate at the reference footprint scale.
+
+    Implemented methods:
+
+        * time-stability filtering
+        * simple averaging
+
+    Parameters
+    ----------
+    ref_class: pygeogrids Grid obj.
+        Class containing the method read_ts for reading the data of the reference
+    others_class: dict
+        Dict of shape 'other_name': pygeogrids Grid obj. for the other dataset
+    geo_subset: tuple, optional. Default is None.
+        Information on the grographic subset for data averaging -> (latmin, latmax, lonmin, lonmax)
+    manager_parms: dict
+        Dict of DataManager attributes
+    """
+
+    def __init__(
+            self,
+            ref_class,
+            others_class,
+            geo_subset,
+            manager_parms,
+    ):
+        self.ref_class = ref_class
+        self.others_class = others_class
+        self.geo_subset = geo_subset
+        # attributes used by the Mixin class:
+        self.datasets = manager_parms["datasets"]
+        self.period = manager_parms["period"]
+        self.read_ts_names = manager_parms["read_ts_names"]
+
+    @property
+    def lut(self) -> dict:
+        """Get a lookup table that combines the points falling under the same reference pixel"""
+        lut = {}
+        for other_name, other_class in self.others_class.items():
+            try:
+                grid = other_class.grid
+            except AttributeError:
+                # when filters are applied, this will be an AdvancedMaskingAdapter
+                grid = other_class.cls.cls.grid
+            other_points = grid.get_bbox_grid_points(
+                *self.geo_subset,
+                both=True,
+            )
+            other_lut = {}
+            # iterate from the side of the non-reference
+            for gpi, lon, lat in zip(other_points[0], other_points[1], other_points[2]):
+                # list all non-ref points under the same ref gpi
+                try:
+                    ref_gpi = self.ref_class.grid.find_nearest_gpi(lon, lat)[0]
+                except AttributeError:
+                    ref_gpi = self.ref_class.cls.cls.grid.find_nearest_gpi(lon, lat)[0]
```

Comment:
Same as above. In general, if you can figure out stuff by a simple check like `hasattr`, always do this instead of `try - except`.